File: Algorithms_in_C/ch_3/circular_linked_list/josephus_problem.py
class Node:
    def __init__(self, key, next=None):
        self.key = key
        self.next = next


def main():
    node = Node(1, None)

    # Capture user input
    N, M = input("Set N and M (example: '9 5': ").split()
    N = int(N)
    M = int(M)

    # Create linked list
    first_node = node
    for x in range(2, N + 1):
        node.next = Node(x)
        node = node.next

    # Make the linked list into a circular linked list
    node.next = first_node

    # Josephus problem
    while node != node.next:
        for _ in range(1, M):
            node = node.next
        print(f"{node.next.key}", end=" ")
        node.next = node.next.next

    print(f"{node.key}")


if __name__ == "__main__":
    main()
        

File: Algorithms_in_C/ch_3/circular_linked_list/josephus_problem.c
#include <stdio.h>
#include <stdlib.h>

struct node {
  int key;
  struct node *next;
}

main() {
  int i, N, M;
  struct node *t, *x;

  // Capture user input
  scanf("%d %d", &N, &M);

  // Creating the linked list of number
  t = (struct node *)malloc(sizeof *t);
  t->key = 1;
  x = t;
  for (i = 2; i <= N; i++) {
    t->next = (struct node *)malloc(sizeof *t);
    t = t->next;
    t->key = i;
  }

  // Pointing the last node to the first node.
  // Makes the linked list circular
  t->next = x;

  // Josephus problem
  while (t != t->next) {
    for (i = 1; i < M; i++)
      t = t->next;
    printf("%d ", t->next->key);
    x = t->next;
    t->next = t->next->next;
    free(x);
  }
  printf("%d\n", t->key);
}
